student.c

```diff
@@ -1,20 +1,42 @@
-#include <assert.h>
 #include "student.h"
 
+#include <assert.h>
+
+static ll_comparator_t compare_gpa(struct ll_node *first,
+                                   struct ll_node *second) {
+    struct student *first_student = LL_ENTRY(first, struct student, node);
+    struct student *second_student = LL_ENTRY(second, struct student, node);
```

Comment:
Good use of these macros!  You could have saved yourself two dereferences by taking advantage of the fact that `LL_ENTRY` expands to a pointer expression, so you could have gotten the GPA's directly:
```c
double first_gpa = LL_ENTRY(first, struct student, node)->gpa;
double second_gpa = LL_ENTRY(second, struct student, node)->gpa;
```
Substituting these doubles into your conditionals would have saved you some memory accesses.  The total performance penalty is likely negligible, but the code would also be a bit cleaner since you could refactor out common expressions.